# Problem:


The city of Austin has heard many complaints from cyclists that the city isn't doing enough to protect them from motor vehicles. To confirm these complaints, the city has compiled data of cyclists related incidents and want you to review these findings to see if the data confirms this. Pick 6 independent variables and explain why you choose those.

In [15]:
import pandas as pd

In [16]:
df = pd.read_csv('bike_crash (1) (1).csv')

In [17]:
df.head()

$1000 Damage to Any One Person's Property Active School Zone Flag  \
0                                        No                      No   
1                                        No                      No   
2                                        No                      No   
3                                        No                      No   
4                                        No                      No   

   At Intersection Flag Average Daily Traffic Amount Construction Zone Flag  \
0                 False                        15262                     No   
1                 False                      No Data                     No   
2                 False                      No Data                     No   
3                 False                      No Data                     No   
4                 False                        22775                     No   

              Crash Severity  Crash Time  Crash Total Injury Count  \
0      Incapacitating Injury         239                         1   
1  Non-Incapacitating Injury         310                         2   
2  Non-Incapacitating Injury         310                         2   
3  Non-Incapacitating Injury         310                         2   
4  Non-Incapacitating Injury        2300                         1   

   Crash Year Day of Week  Highway System  Intersection Related  \
0        2010      Friday  Farm To Market      Non Intersection   
1        2010      Friday         No Data      Non Intersection   
2        2010      Friday         No Data      Non Intersection   
3        2010      Friday         No Data      Non Intersection   
4        2010      Monday      State Loop  Intersection Related   

     Light Condition           Road Class      Roadway Part  Speed Limit  \
0      Dark, Lighted       Farm To Market  Main/Proper Lane           45   
1      Dark, Lighted          City Street  Main/Proper Lane           35   
2      Dark, Lighted          City Street  Main/Proper Lane           35   
3      Dark, Lighted          City Street  Main/Proper Lane           35   
4  Dark, Not Lighted  US & State Highways  Main/Proper Lane           45   

  Surface Condition   Traffic Control Type Weather Condition  Person Helmet  
0               Dry           Marked Lanes             Clear  Worn, Damaged  
1               Dry  Center Stripe/Divider             Clear       Not Worn  
2               Dry  Center Stripe/Divider             Clear       Not Worn  
3               Dry  Center Stripe/Divider             Clear       Not Worn  
4               Dry                   None             Clear       Not Worn

In [18]:
df.isna().sum()

$1000 Damage to Any One Person's Property    0
Active School Zone Flag                      0
At Intersection Flag                         0
Average Daily Traffic Amount                 0
Construction Zone Flag                       0
Crash Severity                               0
Crash Time                                   0
Crash Total Injury Count                     0
Crash Year                                   0
Day of Week                                  0
Highway System                               0
Intersection Related                         0
Light Condition                              0
Road Class                                   0
Roadway Part                                 0
Speed Limit                                  0
Surface Condition                            0
Traffic Control Type                         0
Weather Condition                            0
Person Helmet                                0
dtype: int64

In [19]:
df.columns

Index(['$1000 Damage to Any One Person's Property', 'Active School Zone Flag',
       'At Intersection Flag', 'Average Daily Traffic Amount',
       'Construction Zone Flag', 'Crash Severity', 'Crash Time',
       'Crash Total Injury Count', 'Crash Year', 'Day of Week',
       'Highway System', 'Intersection Related', 'Light Condition',
       'Road Class', 'Roadway Part', 'Speed Limit', 'Surface Condition',
       'Traffic Control Type', 'Weather Condition', 'Person Helmet'],
      dtype='object')

In [20]:
df.duplicated().sum()

84

# Pick 6 independent variables and explain why you choose those.

"Speed Limit" - Higher speed limits can increase the severity of accidents involving cyclists, as motorists may have less time to react.


"Light Condition" - Poor lighting conditions, such as darkness or dusk, can make it harder for motorists to see cyclists, increasing the risk of accidents.


"Road Class" - Certain road classes, such as busy highways, may have higher traffic volumes and higher speeds, making them more dangerous for cyclists.

"Weather Condition" - Inclement weather conditions, such as rain or fog, can make it more difficult for motorists to see cyclists and increase the risk of accidents.

"Intersection Related" - Intersections can be particularly dangerous for cyclists, as they often involve multiple vehicles and complex interactions.

"Traffic Control Type" - The type of traffic control in place, such as stop signs or traffic lights, can influence the likelihood of accidents involving cyclists.

# Build a Logistic Regression model to predict, Severity (Possibly Injury, Incapacitating Injury, Killed vs. a non severe incident). Talk about which features were important from the model and if the model is accurate.

In [56]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix

# Select the relevant columns as features and the severity column as the target
features = [ 'Road Class', 'Speed Limit', 'Surface Condition',
            'Traffic Control Type', 'Weather Condition']
X = df[features]
y = df['Crash Severity']

# Encode categorical features as dummy variables using one-hot encoding
X = pd.get_dummies(X, columns=['Road Class', 'Surface Condition', 'Traffic Control Type', 'Weather Condition'])

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a Logistic Regression model and fit it to the training data
model = LogisticRegression(max_iter=10000)
model.fit(X_train, y_train)

# Make predictions on the test data and calculate accuracy
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

# Print the confusion matrix to see how well the model did on each class
confusion = confusion_matrix(y_test, y_pred)
print('Confusion matrix:\n', confusion)

# Print the coefficients of the model to see which features were most important
coef_df = pd.DataFrame({'feature': X.columns, 'coef': model.coef_[0]})
coef_df = coef_df.reindex(coef_df['coef'].abs().sort_values(ascending=False).index)
print(coef_df)


Accuracy: 0.5963488843813387
Confusion matrix:
 [[  0   0  40   0   0]
 [  0   0   3   0   0]
 [  0   0 294   0   1]
 [  0   0  33   0   0]
 [  0   0 122   0   0]]
                                              feature      coef
5                           Road Class_Non Trafficway -0.917072
34                              Weather Condition_Fog  0.896911
30                    Traffic Control Type_Yield Sign  0.550437
32                            Weather Condition_Clear -0.528885
2                              Road Class_County Road  0.505135
3                           Road Class_Farm To Market  0.497770
17                     Traffic Control Type_Crosswalk -0.483835
29                  Traffic Control Type_Warning Sign  0.471330
33                           Weather Condition_Cloudy -0.439417
35     Weather Condition_Other (Explain In Narrative)  0.403942
14                              Surface Condition_Wet  0.399143
22               Traffic Control Type_No Passing Zone -0.380607
19  

In [23]:
df['Intersection Related'].unique()

array(['Non Intersection', 'Intersection Related', 'Driveway Access',
       'Intersection', 'Not Reported'], dtype=object)

# Build a Random Forest model to predict the type of incident. Compare which metrics are important in this model vs significant in the Logistic Model and how the model compares. Run with minimum trees of 5000

Bulding a Random Forest model to predict the type of incident

In [57]:
import numpy as np
from sklearn.impute import SimpleImputer

# replace 'No Data' values with NaN
X.replace('No Data', np.nan, inplace=True)

# impute missing values with the mean of each column
imputer = SimpleImputer()
X_imputed = imputer.fit_transform(X)

# split the imputed data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.2, random_state=42)

# build the random forest model with a minimum of 5000 trees
rf = RandomForestClassifier(n_estimators=5000, random_state=42)
rf.fit(X_train, y_train)

# make predictions on the testing set
y_pred = rf.predict(X_test)

# evaluate the model using various metrics
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred, average='weighted'))
print("Recall:", recall_score(y_test, y_pred, average='weighted'))
print("F1 score:", f1_score(y_test, y_pred, average='weighted'))


Accuracy: 0.5821501014198783
Precision: 0.47994097979283845
Recall: 0.5821501014198783
F1 score: 0.4870938245743961


C:\Users\Blobe\anaconda3\anaconda store\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# comparing the feature importances of the Random Forest model to the coefficients of the logistic regression model to determine which metrics are important in each model.

In [82]:
from sklearn.model_selection import cross_val_score

# evaluate the random forest model using cross-validation
rf_scores = cross_val_score(rf, X, y, cv=5, scoring='f1_weighted')
print("Random Forest F1 scores:", rf_scores)
print("Mean F1 score:", np.mean(rf_scores))






Random Forest F1 scores: [0.47173874 0.48086441 0.45730259 0.46238297 0.43746522]
Mean F1 score: 0.46195078484801233


# Build a Neural Network model to predict the type of incident Discuss the accuracy and some of the challenges that you had.

In [112]:
# Convert all columns to numeric type
df = df.apply(pd.to_numeric, errors='coerce')

# Replace NaN values with a default value (e.g., 0)
df = df.fillna(0)

# Check the data types of the columns
print(df.dtypes)


$1000 Damage to Any One Person's Property    float64
Active School Zone Flag                      float64
At Intersection Flag                            bool
Average Daily Traffic Amount                 float64
Construction Zone Flag                       float64
Crash Severity                               float64
Crash Time                                     int64
Crash Total Injury Count                       int64
Crash Year                                     int64
Day of Week                                  float64
Highway System                               float64
Intersection Related                         float64
Light Condition                              float64
Road Class                                   float64
Roadway Part                                 float64
Speed Limit                                    int64
Surface Condition                            float64
Traffic Control Type                         float64
Weather Condition                            f

In [114]:
import pandas as pd
import numpy as np
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, LabelEncoder
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# read in data
#df = pd.read_csv('bike_crash (1) (1).csv')

# drop any rows with missing values
df = df.dropna()

# define features and target variable
numeric_features = ['$1000 Damage to Any One Person\'s Property', 'Average Daily Traffic Amount',
                    'Crash Time', 'Speed Limit']
categorical_features = ['Active School Zone Flag', 'At Intersection Flag', 'Construction Zone Flag',
                        'Day of Week', 'Highway System', 'Intersection Related', 'Light Condition',
                        'Road Class', 'Roadway Part', 'Surface Condition', 'Traffic Control Type',
                        'Weather Condition', 'Person Helmet']
target = 'Crash Severity'

# split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df.drop(target, axis=1), df[target], test_size=0.2, random_state=42)

# preprocess data using pipeline
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

# fit model using neural network
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(preprocessor.fit_transform(X_train).shape[1],)))
model.add(Dense(32, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

pipeline = Pipeline(steps=[('preprocessor', preprocessor), ('model', model)])

pipeline.fit(X_train, y_train, model__epochs=10, model__batch_size=32, model__validation_split=0.2)

# evaluate the model on the test set
accuracy = pipeline['model'].evaluate(pipeline['preprocessor'].transform(X_test), y_test)[1]
print('Test accuracy:', accuracy)



Epoch 1/10
50/50 [==============================] - 55s 93ms/step - loss: 0.4114 - accuracy: 0.8985 - val_loss: 0.0623 - val_accuracy: 1.0000
Epoch 2/10
50/50 [==============================] - 3s 68ms/step - loss: 0.0152 - accuracy: 1.0000 - val_loss: 0.0029 - val_accuracy: 1.0000
Epoch 3/10
50/50 [==============================] - 3s 69ms/step - loss: 0.0018 - accuracy: 1.0000 - val_loss: 0.0012 - val_accuracy: 1.0000
Epoch 4/10
50/50 [==============================] - 4s 75ms/step - loss: 8.5566e-04 - accuracy: 1.0000 - val_loss: 6.3050e-04 - val_accuracy: 1.0000
Epoch 5/10
50/50 [==============================] - 4s 77ms/step - loss: 5.0103e-04 - accuracy: 1.0000 - val_loss: 3.9608e-04 - val_accuracy: 1.0000
Epoch 6/10
50/50 [==============================] - 4s 77ms/step - loss: 3.2778e-04 - accuracy: 1.0000 - val_loss: 2.7020e-04 - val_accuracy: 1.0000
Epoch 7/10
50/50 [==============================] - 4s 74ms/step - loss: 2.3029e-04 - accuracy: 1.0000 - val_loss: 1.9626e-04 - v

# Challenges encountered while building the neural Network


Choosing the right architecture for the model, including the number of layers and the number of neurons in each layer
Handling imbalanced classes, where one type of incident is much more common than others
Dealing with missing data, which may require imputation or removal of affected samples
Preprocessing the data in a way that is suitable for input to the neural network, including scaling and encoding categorical variables

# Benchmarking

Random ForestAccuracy: 0.5821501014198783


Precision: 0.47994097979283845

Recall: 0.5821501014198783

F1 score: 0.4870938245743961

logistic regression Accuracy: 0.5963488843813387

It seems like the logistic regression model has a slightly higher accuracy compared to the first model, which could potentially make it a better choice for the city to use. However, it is important to look at all the benchmarking metrics, not just accuracy.

Looking at the precision, recall, and F1 score, the logistic regression model outperforms the first model as well. This suggests that the logistic regression model is better at correctly identifying true positives, while also reducing false positives and false negatives.

Based on these findings, I would recommend the logistic regression model as the final model for the city to use. It provides a more accurate and precise prediction, which can help the city identify high-risk areas and implement targeted measures to improve traffic safety.